backend/api/EventHandlers/MissionEventHandler.cs

```diff
@@ -102,111 +108,103 @@ private async void OnRobotAvailable(object? sender, RobotAvailableEventArgs e)
                 return;
             }
 
-            if (MissionRunQueueIsEmpty(MissionRunQueue(robot.Id)))
+            if (MissionScheduling.MissionRunQueueIsEmpty(MissionRunQueue(robot.Id)))
             {
                 _logger.LogInformation("The robot was changed to available but there are no mission runs in the queue to be scheduled");
                 return;
             }
 
-            var missionRun = MissionRunQueue(robot.Id).First(missionRun => missionRun.Robot.Id == robot.Id);
+            var missionRun = MissionRunQueue(robot.Id).FirstOrDefault(missionRun => missionRun.Robot.Id == robot.Id);
+
+            if (robot.MissionQueueFrozen == true)
+            {
+                missionRun = MissionRunQueue(robot.Id).FirstOrDefault(missionRun => missionRun.Robot.Id == robot.Id &&
+                        missionRun.MissionRunPriority == MissionRunPriority.Emergency);
+
+                if (missionRun == null)
+                {
+                    _logger.LogInformation("The robot was changed to available in emergency state and no emergency mission run is scheduled");
+                    return;
+                }
+            }
+
+            if (missionRun == null)
+            {
+                _logger.LogInformation("The robot was changed to available but no mission is scheduled");
+                return;
+            }
 
             _scheduleMissionMutex.WaitOne();
-            StartMissionRunIfSystemIsAvailable(missionRun);
+            MissionSchedulingService.StartMissionRunIfSystemIsAvailable(missionRun);
             _scheduleMissionMutex.ReleaseMutex();
         }
 
-        private void StartMissionRunIfSystemIsAvailable(MissionRun missionRun)
+        private async void OnEmergencyButtonPressedForRobot(object? sender, EmergencyButtonPressedForRobotEventArgs e)
         {
-            if (!TheSystemIsAvailableToRunAMission(missionRun.Robot, missionRun).Result)
+            _logger.LogInformation("Triggered EmergencyButtonPressed event for robot ID: {RobotId}", e.RobotId);
+            var robot = await RobotService.ReadById(e.RobotId);
+            if (robot == null)
+            {
+                _logger.LogError("Robot with ID: {RobotId} was not found in the database", e.RobotId);
+                return;
+            }
+
+            var area = await AreaService.ReadById(e.AreaId);
+            if (area == null)
             {
-                _logger.LogInformation("Mission {MissionRunId} was put on the queue as the system may not start a mission now", missionRun.Id);
+                _logger.LogError("Could not find area with ID {AreaId}", e.AreaId);
                 return;
             }
 
+            await MissionSchedulingService.FreezeMissionRunQueueForRobot(e.RobotId);
```

Comment:
In this function we should either check to make sure the robot is not null, or we need to put this function call inside the try catch

Reply:
We do this on line 146, is that what you were asking about?

Reply:
sorry, I missed that. I guess in theory the state could change in the mean time, but it's unlikely.